content/posts/2024-12-04-improve-clang-doc.md

```diff
@@ -0,0 +1,125 @@
+---
+author: "Peter Chou"
+date: "2024-12-04"
+tags: ["GSoC", "Clang-Doc"]
+title: "GSoC 2024: Improve Clang Doc"
+---
+
+Hi, my name is Peter, and this year I was involved in Google Summer of Code 2024. I worked on [improving the Clang-Doc documenation generator](https://discourse.llvm.org/t/improve-Clang-Doc-usability/76996)
+
+Mentors: Petr Hosek and Paul Kirth
+
+## Project Background
+
+Clang-Doc is a documentation generator developed on top of libtooling, as an
+alternative to Doxygen. Development started in 2018 and continued through 2019,
+however, it has since stalled. Currently, the tool can generate HTML, YAML, and markdown but the generated output has usability issues. This GSOC project aimed to address the pain points regarding the output of the HTML, by adding support for various C++ constructs and reworking the CSS of the HTML output to be more user-friendly.
+
+## Work Done
+
+The original scope of the project was to improve the output of Clang-Doc's generation. However during testing the tool was significantly slower than expected which made developing features for the tool impossible. 
+Each compilation of the LLVM codebase was taking upwards of 10 hours on my local machine. Additionally, the tool utilized a lot of memory and was prone to crashing with an out-of-memory error. Similar tools such as Doxygen and Hdoc ran in comparatively less time for the same codebase. This pointed to a significant bottleneck within Clang-Doc’s code path when generating large-scale software projects. Due to this the project scope quickly changed to improving the runtime of Clang-Doc so that it could run much faster. It was only during the latter half of the project did the scope changed back to improving Clang-Doc’s generation.
+
+### Added More Test Cases to Clang-Doc test suite
+
+Clang-Doc previously had tests which did not test the full scope of the the HTML or Markdown output. I added more end-to-end tests to make sure that in the process of optimizing documentation generation we were not degrading the quality or functionality of the tool. 
+
+In summary, I added four comprehensive tests that covered all features that we were not testing such as testing the generation for Enums, Namespace, and Records for HTML and Markdown.
+
+### Improve Clang-Doc’s performance by 1.58 times 
+
+Internally, the way Clang-Doc works is by leveraging libtooling's ASTVisitor class to parse the source level declarations in each TU. 
+
+Clang-Doc is architected using a Map-Reduce pattern. Clang-Doc parses each fragment of a declaration into an in-memory data format which is serialized then into an internal format and stored as a key value paired, identified by their [USR](https://clang.llvm.org/doxygen/group__CINDEX__CURSOR__XREF.html#ga51679cb755bbd94cc5e9476c685f2df3). After, Clang-Doc deserializes and combines each of the fragment declarations back into the in-memory data format which is used by each of the backend to generate the results.
```

Comment:
Nit: every sentence shouldn't start w/ `Clang-Doc`. `It' may be appropriate for some, but you could equally rephrase as `we` or use another literary construct altogether. The point is, it reads as very repetitive.